src/ucp/core/ucp_worker.c

```diff
@@ -750,18 +753,152 @@ void ucp_worker_iface_event(int fd, void *arg)
+        if ((ctx->tl_rscs[id].dev_index != ctx->tl_rscs[wiface->rsc_index].dev_index) ||
+            (if_iter->flags & UCP_WORKER_IFACE_FLAG_SUBOPTIMAL) ||
+            (id == wiface->rsc_index)) {
+            continue;
+        }
+
+        /* Check that another iface:
+         * 1. Supports all capabilities of the target iface (at least)
+         * 2. Has the same or better performance charasteristics */
+        if (ucs_test_all_flags(if_iter->attr.cap.flags, wiface->attr.cap.flags) &&
+            (if_iter->attr.overhead          <= wiface->attr.overhead) &&
+            (if_iter->attr.latency.overhead  <= wiface->attr.latency.overhead) &&
+            (if_iter->attr.latency.growth    <= wiface->attr.latency.growth) &&
+            (if_iter->attr.bandwidth         >= wiface->attr.bandwidth) &&
+            (if_iter->attr.priority          >= wiface->attr.priority)) {
+
+            /* Do not check this iface anymore, because better one exists.
+             * It helps to avoid the case when two interfaces with the same caps
+             * and performance exclude each other. */
+            wiface->flags |= UCP_WORKER_IFACE_FLAG_SUBOPTIMAL;
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+/**
+ * @brief Find the minimal possible set of tl interfaces for each device
+ *
+ * @param [in]  worker     UCP worker.
+ * @param [out] tl_bitmap  Map of the relevant tl resources.
+ *
+ * @return Error code as defined by @ref ucs_status_t
+ */
+static ucs_status_t ucp_worker_select_best_ifaces(ucp_worker_h worker,
+                                                  uint64_t *tl_bitmap_p)
+{
+    ucp_context_h context = worker->context;
+    uint64_t tl_bitmap    = 0;
+    ucp_worker_iface_t *ifaces;
+    ucp_worker_iface_t *wiface;
+    ucp_rsc_index_t tl_id, iface_id;
+    unsigned num_ifaces;
+
+    /* For each iface check whether there is another iface, which:
+     * 1. Supports at least the same capabilities
+     * 2. Provides equivalent or better performance
+     */
+    for (tl_id = 0; tl_id < context->num_tls; ++tl_id) {
+        wiface = &worker->ifaces[tl_id];
+        if (!ucp_worker_iface_find_better(worker, wiface)) {
+            tl_bitmap |= UCS_BIT(tl_id);
+        }
+    }
+
+    *tl_bitmap_p = tl_bitmap;
+    num_ifaces   = ucs_count_one_bits(tl_bitmap);
+    ucs_assert(num_ifaces <= context->num_tls);
+
+    if (num_ifaces < context->num_tls) {
+        /* Some ifaces need to be closed */
+        ifaces = ucs_calloc(num_ifaces, sizeof(ucp_worker_iface_t), "ucp iface");
+        if (ifaces == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        iface_id = 0;
+        for (tl_id = 0; tl_id < context->num_tls; ++tl_id) {
+            wiface = &worker->ifaces[tl_id];
+            if (tl_bitmap & UCS_BIT(tl_id)) {
+                memcpy(ifaces + iface_id, wiface, sizeof(*wiface));
+                ++iface_id;
+            } else {
+                ucs_debug("closing suboptimal interface[%d]=%p on "
```

Comment:
maybe print which resource replaces which
e.g `closing resource[0] rc/mlx5_0:1 since resource[1] rc/mlx5_0:1 is better`